src/menu_events.cpp

```diff
@@ -649,13 +649,14 @@ void menu_handler::rename_unit()
 unit_map::iterator menu_handler::current_unit()
 {
 	const mouse_handler& mousehandler = pc_.get_mouse_handler_base();
+	const bool see_all = pc_.is_replay() && dynamic_cast<playsingle_controller&>(pc_).get_replay_controller()->see_all();
```

Comment:
Is this `dynamic_cast` good? Should I add a `virtual replay_controller * play_controller::get_replay_controller() { return nullptr; }` and have `playsingle_controller::get_replay_controller` override it?

Reply:
Yes, I think you should move `get_replay_controller()` up to `play_controller`. I'd also move the `replay_controller` member up so you can make sure in code that if `is_replay()` is true there is a valid `replay_controller`.

Reply:
I've moved get_replay_controller up. I didn't understand what you mean by moving the data member, though. Is the code good as it is now or should I make further changes?

Reply:
Using `get_replay_controller()` in `play_controller::is_replay()` like you did works as well. Now you can remove `playsingle_controller::is_replay()` since it does the same thing.

Reply:
I already have, haven't I? Thanks for the review.

Reply:
Ah, indeed I missed that, my bad.